pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,321 @@
+func NewClient(c Connection, encryptor Encryptor, decrypter Decryptor) (*Client, error) {
+	client := &Client{
+		encryptor: encryptor,
+		decryptor: decrypter,
+	}
+	if c != nil {
+		client.conn = c
+		return client, nil
+	}
+	err := client.NewConnection()
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// Prepare prepares the given string into a sql statement on the client's connection.
+func (c *Client) Prepare(stmt string) *sql.Stmt {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	prepared, err := c.conn.Prepare(stmt)
+	if err != nil {
+		panic(errors.Errorf("Error preparing statement: %s\n%v", stmt, err))
+	}
+	return prepared
+}
+
+// QueryForRows queries the given stmt with the given params and returns the resulting rows. The query wil be retried
+// given a sqlite busy error.
+func (c *Client) QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	var rows *sql.Rows
+	var err error
+
+	err = wait.ExponentialBackoff(backoffRetry, func() (bool, error) {
+		rows, err = stmt.QueryContext(ctx, params...)
+		if err != nil {
+			sqlErr, ok := err.(*sqlite.Error)
+			if ok && sqlErr.Code() == sqlite3.SQLITE_BUSY {
+				return false, nil
+			}
+			return false, errors.Wrapf(err, "Error initializing Store DB")
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rows, nil
+}
+
+// CloseStmt will call close on the given Closable. It is intended to be used with a sql statement. This function is meant
+// to replace stmt.Close which can cause panics when callers unit-test since there usually is no real underlying connection.
+func (c *Client) CloseStmt(closable Closable) error {
+	return closable.Close()
+}
+
+// ReadObjects Scans the given rows, performs any necessary decryption, converts the data to objects of the given type,
+// and returns a slice of those objects.
+func (c *Client) ReadObjects(rows Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	// defer rows.Close()
+
+	// rt := time.Now()
+	var result []any
+	for rows.Next() {
+		data, err := c.decryptScan(rows, shouldDecrypt)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		singleResult, err := fromBytes(data, typ)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		result = append(result, singleResult.Elem().Interface())
+	}
+	err := rows.Err()
+	if err != nil {
+		return nil, closeRowsOnError(rows, err)
+	}
+
+	err = rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// ReadStrings scans the given rows into strings, and then returns the strings as a slice.
+func (c *Client) ReadStrings(rows Rows) ([]string, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+
+	var result []string
+	for rows.Next() {
+		var key string
+		err := rows.Scan(&key)
+		if err != nil {
+			ce := rows.Close()
```

Comment:
Think you can do `return nil, closeRowsOnError(rows, err)` here

And on that subject, I would  have `closeRowsOnError` return `nil, ERROR` so you can just do `return closeRowsOnError(rows, err)`

Reply:
Replace occurences of these with closeRowsOnError. I can have it return nil because I call it in `ReadObjects` which has a different return type. Updated function comment for closeRowsOnError as well, name was wrong.